[distributed] cleanup dist autograd context on other nodes when it is released on one node

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27951 [distributed] cleanup dist autograd context on other nodes when it is released on one node**

Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

In follow up PRs, we will add error checking + retries for this call.
Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

rohan-varma committed:
[wip][distributed] send rpc for releaseContext

Sends an rpc for release context

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] send rpc to release dist autograd context on other nodes"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes will not have their contexts released. In follow-up PRs, we can implement bettter cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] send rpc to release dist autograd context on other nodes"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes will not have their contexts released. In follow-up PRs, we can implement bettter cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] send rpc to release dist autograd context on other nodes"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes will not have their contexts released. In follow-up PRs, we can implement bettter cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] send rpc to release dist autograd context on other nodes"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes will not have their contexts released. In follow-up PRs, we can implement bettter cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] send rpc to release dist autograd context on other nodes"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes will not have their contexts released. In follow-up PRs, we can implement bettter cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the current context and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying.
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 I think this is ready for a first pass look now. some legit failures (https://circleci.com/gh/pytorch/pytorch/3209870?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) in dist autograd tests since the contexts are being cleaned up across nodes now, so the tests might need some more refactoring.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -195,6 +197,18 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::processRpc(
+      // if it exists since it may have been deleted by an in-flight RPC.
+      if (DistAutogradContainer::getInstance().hasContextWithId(
+              cleanupContextId)) {
+        DistAutogradContainer::getInstance().releaseContext(
```
Actually, we probably need to do this check + release atomically, since it is still possible that another thread could release the context after we've done the `hasContextWithId` check. Then, `releaseContext` would throw but we don't necessarily want to throw, since the context has been freed as desired. Should we make a version of releaseContext that does not throw if the context we are releasing does not exist (or control this with a param?)

pritamdamania87 replied:
Yes, let's do check + release automatically and have a version which doesn't throw if the context is cleaned up (can call it releaseContextIfPresent).

pritamdamania87 replied:
Also, shouldn't `notifyWorkers` be true here? In case we do nested RPCs, we need this node to notify other nodes too.

rohan-varma replied:
It should eventually be true, but since this PR doesn't implement cycle detection this is a way of not doing additional RPC calls. Once #27576 lands we can modify this to support nested rpc calls.

pritamdamania87 replied:
Do we need to do cycle detection? If we do an atomic check + release, we might ask a node to release again, but that would just be a noop right?

rohan-varma replied:
Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

pritamdamania87 replied:
> Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

True, although for now let's just go with a simple version and later on we can evaluate whether we need to optimize this further.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -13,6 +13,11 @@
+
+def store_context_id(context_id):
+    global known_context_ids
+    known_context_ids.append(context_id)
```
Add a small comment here mentioning we don't need a lock here since we hold the GIL when executing python UDFs.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -149,10 +154,8 @@ def test_autograd_functions(self):
-                time.sleep(0.1)
-                pass
+            # Synchronize all workers.
+            rpc.sync_rpc()
```
Lets use `dist.barrier()` since `sync_rpc` might go away in a few future changes we have planned.

rohan-varma replied:
I don't think `dist.barrier()` works here (just tried it and the tests fail if we replace it). It appears that `dist.barrier()` blocks until all the processes enter the barrier. But `rpc.sync_rpc()` gives us the further guarantee that all outstanding RPCs will be completed, which is what we need here, since we need the RPCs for releasing the context to finish. With `dist.barrier()` there is no guarantee that all outstanding RPCs will be completed, it looks like. cc @mrshenli in case my understanding isn't correct

pritamdamania87 replied:
Hmm what is the reason `dist.barrier` doesn't work here? We are only doing sync RPCs in this test so all RPCs should have finished on all nodes when we are done with the barrier.

> Why is sync_rpc going away? Is there going to be any other way to enforce all processes finish all outstanding RPCs and then move forward (without joining)?

Because the user should wait on RPCs that they care about and not the framework. If you care about an RPC, you should use future.wait() to check its result.

rohan-varma replied:
But the RPC that is sent to release the context on other nodes isn't synchronous right? It is sent with `agent->send()` which enqueues the send, and the user doesn't wait on this (i.e. when the context manager is exited, the RPCs are sent, but we don't wait on these RPCs). If we want, we can wait on the future inside `releaseContext`, but that would be a slowdown to the user.

rohan-varma replied:
In this particular case `dist.barrier()` works since the async RPCs haven't been sent, but in the other test where we sync after exiting the context manager, we would need a way to make sure the RPCs are sent + processed before continuing with the test.

rohan-varma replied:
> Because the user should wait on RPCs that they care about and not the framework.

Agreed. What about the case where our framework itself sends async RPCs that as part of handling a user request? This is basically what these cleanup context RPCs are. The user does not have any handle to the futures created by them, but we should at some point guarantee that they are done.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -220,6 +223,28 @@ def test_rpc_complex_args(self):
+                ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
+                rpc.rpc_sync("worker{}".format(dst_rank), store_context_id, args=(context_id,))
+        # now let each worker finish with their cleanup RPCs
+        rpc.sync_rpc()
```
dist.barrier().

rohan-varma replied:
See above conversation regarding this

pritamdamania87 replied:
One way to fix this would be to have loop with a timeout which checks for the context and sleeps for a while if the context still exists. If we exit this loop and the timeout has expired, we fail the test. The timeout could be 10s to be conservative. An example:

```
start = time.time()
bool success = False
while time.time() - start < 10:
  if not dist_autograd.has_context_id(context_id):
    success = True
    break
self.assertTrue(success)
```

This structure can also be moved into a helper function that we can reuse in other tests.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <torch/csrc/distributed/autograd/context/dist_autograd_container.h>
 #include <c10/util/Exception.h>
+#include <torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.h>
```
nit: Try to order includes alphabetically. I thought clang-tidy would catch this? Maybe it doesn't?

rohan-varma replied:
This is strange, when I run the auto linter it gives me this ordering. If I change the include order there is a linter error.

pietern replied:
There must be a newline between the top include (for the header corresponding to the source file) and the other includes.

rohan-varma replied:
I'll clean this up (and the other files that this is missing) as part of a later diff. Should we add a lint rule for this?

pietern replied:
It's a clang-format rule, not a clang-tidy rule. We don't lint on clang-format yet (see #25700).

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -113,6 +114,11 @@ bool DistAutogradContainer::hasValidContext() const {
   return current_context_id_ != kInvalidContextId;
 }
 
+bool DistAutogradContainer::hasContextWithId(int64_t context_id) const {
```
nit: `hasContextId`

rohan-varma replied:
This will actually go away since we are checking + releasing atomically with `releaseContextIfPresent`

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -128,12 +134,27 @@ DistAutogradContext& DistAutogradContainer::currentContext() {
-void DistAutogradContainer::releaseContext(int64_t context_id) {
+void DistAutogradContainer::releaseContext(
+    int64_t context_id,
+    bool notifyWorkers) {
```
In which case would `notifyWorkers` be false?

rohan-varma replied:
Right now, this is here since this PR doesn't support cycle detection, so this is a simple flag to avoid cycles. In a follow up PR I was planning to add additional changes to supported nested RPCs. 

More generally, there may be cases where we know that we don't want to identify additional workers (for example if we know all of them have already cleaned up the context), then we can avoid extra RPCs with this flag.

rohan-varma replied:
Since we now have `releaseContextIfPresent`, `notifyWorkers` should no longer be needed since the RPC would just do a no-op if the context didn't exist, so we can get rid of `notifyWorkers`.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -128,12 +134,27 @@ DistAutogradContext& DistAutogradContainer::currentContext() {
+        autograd_context_.find(context_id)->second.getKnownWorkerIds();
+    auto agent = rpc::RpcAgent::getDefaultRpcAgent();
+    for (const auto& worker_id : workerIds) {
+      agent->send(
```
Create a gh issue mentioning that we should store the futures we get from these calls in a structure and retry them later if they failed.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -220,6 +223,28 @@ def test_rpc_complex_args(self):
+            t1 = torch.ones(3, 3, requires_grad=True)
+            t2 = torch.zeros(3, 3, requires_grad=True)
+            for dst_rank in dst_ranks:
+                ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
```
Create a gh issue that we should add a test with multiple nested RPCs (with Python UDFs) once https://github.com/pytorch/pytorch/pull/27576 lands.

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp:
```diff
@@ -0,0 +1,46 @@
+
+rpc::Message CleanupAutogradContextReq::toMessage() && {
+  std::vector<at::IValue> ivalues;
+  ivalues.emplace_back(context_id_);
```
nit: `std::vector<at::IValue> ivalues {context_id_};`

rohan-varma replied:
No longer needed per your suggestion since we just have a single value we are pickling.

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp:
```diff
@@ -0,0 +1,46 @@
+  // pickle
+  std::vector<torch::Tensor> tensorTable;
+  std::vector<char> payload =
+      jit::pickle(c10::ivalue::Tuple::create(std::move(ivalues)), &tensorTable);
```
Why do we need to create a tuple here? We just have one IValue, can't we just pickle that?

rohan-varma replied:
I was thinking of keeping it this way since we could potentially avoid a refactor if we decide to start sending additional data for this RPC.

pritamdamania87 replied:
For now, let's avoid using a tuple since it would have some overhead. If later on we need this we can change this accordingly.

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_resp.h:
```diff
@@ -0,0 +1,23 @@
+namespace distributed {
+namespace autograd {
+
+// Used to request other workers to clean up their autograd context.
```
Mention that this is an empty response for the `CleanupAutogradContextReq`. The current comment seems to indicate this call is responsible for cleaning up the context.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/autograd/context/dist_autograd_container.h:
```diff
@@ -73,6 +79,15 @@ class TORCH_API DistAutogradContainer {
 
+  // Sends an RPC to the workers that have a context corresponding to passed in
+  // context_id. This function should be called with the lock.
+  void sendReleaseContextRpc(int64_t context_id);
```
These methods are added to reduce code duplication between `releaseContext` and `releaseContextIfPresent`.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -202,11 +202,10 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::processRpc(
-        DistAutogradContainer::getInstance().releaseContext(
-            cleanupContextId, false);
-      }
+      // This can created nested RPCs if there are other nodes that get notified
```
nit: s/created/create

zhaojuanmao commented on test/dist_autograd_test.py:
```diff
@@ -13,6 +14,13 @@
+    known_context_ids.append(context_id)
 
 def _set_rpc_done(context_id):
     global prev_rank_rpc_done
```
it can be deleted since it is not used anymore

zhaojuanmao commented on torch/csrc/distributed/autograd/context/dist_autograd_container.h:
```diff
@@ -36,8 +36,15 @@ class TORCH_API DistAutogradContainer {
+  // context. Does nothing if it is not present.
+  void releaseContextIfPresent(int64_t context_id);
+
   // Retrieve the autograd context for a given context_id.
```
wondering why they can not be merged?

rohan-varma replied:
One checks if the context exists, one throws if it does not. We can merge them if we have a flag `throwIfNotExists` or something similar, should we do this? cc @pritamdamania87

pritamdamania87 replied:
Two separate methods seems a bit cleaner to me than having a flag.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented on test/dist_autograd_test.py:
```diff
@@ -4,22 +4,46 @@
+# after dist autograd context is cleaned up, it should be cleaned up on other
+# nodes. This helper allows timeout_seconds for those RPCs to be completed, and
+# ensures that all the contexts have been cleaned up in that timeframe.any
+def _all_contexts_cleaned_up(num_contexts, timeout_seconds=10):
```
@pritamdamania87 I implemented your suggestion and removed the use of `rpc.sync_rpc()` - is this 10s timeout ok or will we need a longer one for TSAN/ASAN tests?

pritamdamania87 replied:
10s should be sufficient.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.h:
```diff
@@ -36,8 +36,15 @@ class TORCH_API DistAutogradContainer {
 
   // Clean up resources for a given context_id once the autograd pass is done.
+  // Sends RPC to other workers this worker knows about, telling them to clean
+  // up their context as well.
```
Lets mention that this throws an exception if context_id not found.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented:
Waiting on #28312 to land before landing this, since that is more urgent and it would be delayed by required rebase/test fixes if this went in first.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

Note: the current version is very simple and does not attempt to do cycle detection of RPCs or any retries. If `releaseContext` is called directly on one node, then that node will send RPCs to the other nodes it knows about to release their context. However, if a node receives an RPC that tells it to release its context, it will do so, but will not forward this request to other nodes that it knows about. This is to avoid cycles. The limitation of this approach is that the entire graph of nodes may not have their contexts released. In follow-up PRs, we can implement better cycle detection to solve this problem.

Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma commented on test/dist_autograd_test.py:
```diff
@@ -258,6 +285,9 @@ def _test_graph(self, fn):
             self._verify_graph_for_rpc_call_exec(list(send_functions.values())[0])
+            # this barrier is needed so one worker does not clean up their
+            # autograd context before another worker tries to access it.
+            dist.barrier()
```
@zhaojuanmao with these changes, had to add `dist.barrier()`s here and in other tests so there is no races for the autograd context.

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

In follow up PRs, we will add error checking + retries for this call.
Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

In follow up PRs, we will add error checking + retries for this call.
Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] cleanup dist autograd context on other nodes when it is released on one node"


Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

In follow up PRs, we will add error checking + retries for this call.
Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

[ghstack-poisoned]
